packages/net/tcp_connection.pony

```diff
@@ -495,8 +496,15 @@ actor TCPConnection
             _notify_connecting()
           end
         else
-          // We're already connected, unsubscribe the event and close.
-          @pony_asio_event_unsubscribe(event)
+          // There is a possibility that a non-Windows system has
+          // already unsubscribed this event already.  (Windows might
+          // be vulnerable to this race, too, I'm not sure.) It's a
+          // bug to do a second time.  Look at the disposable status
+          // of the event (not the flags that this behavior's args!)
+          // to see if it's ok to unsubscribe.
+          if not @pony_asio_event_get_disposable(event) then
+            @pony_asio_event_unsubscribe(event)
+          end
           @pony_os_socket_close[None](fd)
```

Comment:
I haven't looked closely enough at this, but should this `pony_os_socket_close` also only be done if the event is `not disposable`? I believe the normal code path that calls `pony_asio_event_unsubscribe` also closes the relevant `fd` and it would be an error if we tried closing it again (although `pony_os_socket_close` doesn't check for errors so we wouldn't notice either way).

Reply:
I used to include both unsubscribe & close calls in the clause, but then I moved the close out again.  The Pony runtime doesn't have an assert inside `@pony_os_socket_close()` for a `-1` file descriptor, and the OS doesn't care either (in the sense that the system call fails but nothing terrible happens to user- or kernel-space afterward).

Reply:
Works for me.